### PR TITLE
Improved ffi support for LuaJIT, +269% perf boost

### PR DESCRIPTION
**Pull request by ExtReMLapin**
43350% performance boost on lei2str

Didn't benchmark str2lei but it's also faster

Edit:

after all the optimizations done here, it's a 

* **+269% perf boost on LuaJIT**
* +5% perf boost on LuaPUC 5.1

### CODE REVIEW

**Comment by tst2005** [comment]
Hello If i'm not wrong your changes introduce "ffi" as mandatory required.
But md5.lua is supposed to be run with or without "ffi" : PUC/Rio LUA don't have "ffi".
Please consider the case of "ffi" not available : fallback to a code without "ffi".
Regards,

**Comment by ExtReMLapin** [comment]
yes, my bad, I forgot to push the very last version of my edits

**Comment by ExtReMLapin** [comment]
As a performance PoV it would be probably better to define the function inside the condition and not have the condition inside the function, but whatever.

**Comment by ExtReMLapin** [comment]
200% perf boost from 880eaf24c408e2f1ec6102f7d09bdef5ad821069 to 50266aca00e375b19a4a1acefc7230945434e131

File: md5.lua [hunk 1]
```diff
@@ -347,7 +365,7 @@ local function md5_update(self, s)
   self.pos = self.pos + #s
   s = self.buf .. s
   for ii = 1, #s - 63, 64 do
-    local X = cut_le_str(sub(s,ii,ii+63),4,4,4,4,4,4,4,4,4,4,4,4,4,4,4,4)
```
Comment (ExtReMLapin) [inline md5.lua:350]: Why in the hell was this even here ???

**Comment by pablomayobre** [comment]
This is really good, I'd approve this change!

**Review by kikito — APPROVED** [review]

**Comment by kikito** [comment]
Thanks for this, I have merged it. Apologies for the time it took me to do it.

**Pull request merged**